Проблема с аутентификацией?

У меня не получилось сбилдить куни с UI, поэтому его вырезал. Дошел до входа в акк и получается вот так:
```
[02:28:30][UI thread][Logger][INFO]: Log file: /tmp/aui.97846.log
[02:28:30][UI thread][App][INFO]: Waiting for Telegram network...
[02:28:32][UI thread][TelegramClient][INFO]: [Authentication] required. Please supply phone number to stdin
+1860734XXXX
[02:28:40][UI thread][Performance][WARN]: Execution of a task took 8297753us to execute which may cause UI lag.
 - ...

[02:28:40][UI thread][App][INFO]: Bot is shutting down. Please give some time to dump remaining context
```
Моментально после ввода номера идут последующие сообщения и приложуха падает.
Подскажите плз в чем проблема, спасибо заранее

Логи билда я уже убрал, но было куча ошибок вроде no matching function for call to ‘AText::AText(AString) и DiaryQueryAI.cpp:110:58: error: no match for ‘operator&’ (operand types are ‘AVerticalLayout’ and ‘ass::PropertyListRecursive’). И лились сотни строк говна про деривации, не читал

> Логи билда я уже убрал, но было куча ошибок вроде no matching function for call to ‘AText::AText(AString) и DiaryQueryAI.cpp:110:58: error: no match for ‘operator&’ (operand types are ‘AVerticalLayout’ and ‘ass::PropertyListRecursive’). И лились сотни строк говна про деривации, не читал

а, ну да. я эту херню не закоммитил в ауй

--- a/src/AppBase.cpp
+++ b/src/AppBase.cpp
@@ -36,6 +36,7 @@ static const auto WORKING_MEMORY_PATH = APath("data") / "working_memory.md";
 
 
 AFuture<std::valarray<double>> contextEmbedding(ranges::range auto && rng) {
+    ALOG_TRACE(LOG_TAG) << "contextEmbedding";
     AString basePrompt;
     AUI_ASSERT(!ranges::empty(rng));
     for (const OpenAIChat::Message& message: rng) {
@@ -327,12 +328,14 @@ AppBase::AppBase(APath workingDir): mDiary(workingDir / "diary"), mWakeupTimer(_
 }
 
 const AppBase::Notification& AppBase::passNotificationToAI(AString notification, OpenAITools actions, bool first) {
+    ALOG_TRACE(LOG_TAG) << "passNotificationToAI";
     const auto& result = *mNotifications.emplace(first ? mNotifications.begin() : mNotifications.end(), std::move(notification), std::move(actions));
     mNotificationsSignal.supplyValue();
     return result;
 }
 
 AFuture<> AppBase::diaryDumpMessages() {
+    ALOG_TRACE(LOG_TAG) << "diaryDumpMessages";
     // mDiary.reload(); // will find plagiarism against all entries. // commented out: exclude plagiarism checks for
     // included entries
     AUI_DEFER { mDiary.reload(); };
@@ -403,6 +406,7 @@ AFuture<> AppBase::diaryDumpMessages() {
 }
 
 void AppBase::actProactively() {
+    ALOG_TRACE(LOG_TAG) << "actProactively";
     AString prompt = "<your_diary_page just_for_reasoning no_plagiarism no_copy>\n";
     if (!mDiary.list().empty()) {
         auto idx = re() % mDiary.list().size();
@@ -434,6 +438,7 @@ Act proactively!
 
 
 void AppBase::updateTools(OpenAITools& actions) {
+    ALOG_TRACE(LOG_TAG) << "updateTools";
     actions.insert({
         .name = "ask_diary",
         .description = "Consult with Kuni's main knowledge database (subagent). Use this to retrieve additional pages from diary. USE THIS PROACTIVELY — especially when someone shares personal news, asks about past events, or mentions people/activities you might know about.\n\nExamples of when to call:\n- User says \"I wrote a song today\" → query: \"[sender name] said they wrote a song today. What do I know about them and songs? Do they participate in a band? Which songs do they write? What music do they listen to?\"\n- User asks \"what songs am I writing?\" → query: \"What songs does [sender name] write? What do I know about their musical activities?\"\n- User says \"I'm going to the gym\" → query: \"Does [sender name] go to the gym? Any related habits or routines?\"",
@@ -496,12 +501,14 @@ void AppBase::updateTools(OpenAITools& actions) {
 }
 
 void AppBase::removeNotifications(const AString& substring) {
+    ALOG_TRACE(LOG_TAG) << "removeNotifications: " << substring;
     mNotifications.erase(ranges::remove_if(mNotifications, [&](const Notification& n) {
         return n.message.contains(substring);
     }), mNotifications.end());
 }
 
 AString AppBase::takeDiaryEntry(const Diary::EntryExAndRelatedness& i) {
+    ALOG_TRACE(LOG_TAG) << "takeDiaryEntry: " << i.entry->id;
     if (ranges::any_of(mTemporaryContext, [&](const OpenAIChat::Message& m) {
         return m.content.contains(i.entry->freeformBody);
     })) {

--- a/src/Diary.cpp
+++ b/src/Diary.cpp
@@ -20,12 +20,16 @@
 
 using namespace std::chrono_literals;
 
+static constexpr auto LOG_TAG = "Diary";
+
 Diary::Diary(APath diaryDir)
     : mDiaryDir(std::move(diaryDir)) {
+    ALOG_TRACE(LOG_TAG) << "Diary::Diary: " << diaryDir;
     mDiaryDir.makeDirs();
 }
 
 AVector<Diary::Entry> Diary::read(const APath& diaryDir) {
+    ALOG_TRACE(LOG_TAG) << "Diary::read: " << diaryDir;
     AVector<Entry> result;
     diaryDir.makeDirs();
     for (const auto& entry : diaryDir.listDir()) {
@@ -41,10 +45,12 @@ AVector<Diary::Entry> Diary::read(const APath& diaryDir) {
 }
 
 void Diary::save(const Entry& entry) {
+    ALOG_TRACE(LOG_TAG) << "Diary::save: " << entry.id;
     AFileOutputStream(mDiaryDir / (entry.id + ".md")) << entry.text;
 }
 
 void Diary::save(const EntryEx& entry) {
+    ALOG_TRACE(LOG_TAG) << "Diary::save(EntryEx): " << entry.id;
     // inject
     // ---
     // { ... }
@@ -58,6 +64,7 @@ void Diary::save(const EntryEx& entry) {
 }
 
 AFuture<AVector<Diary::EntryExAndRelatedness>> Diary::query(const std::valarray<double>& query, QueryOpts opts) {
+    ALOG_TRACE(LOG_TAG) << "Diary::query";
     struct DiaryEntryExAndRelatednessF {
         std::list<EntryEx>::iterator entry;
         AFuture<double> relatedness;

--- a/src/Diary.h
+++ b/src/Diary.h
@@ -208,7 +208,7 @@ public:
      * @brief Invalidate the cached diary so that it will be reloaded on
      *        next access.
      */
-    void reload() { mCachedDiary.reset(); }
+    void reload() { ALOG_TRACE("Diary") << "reload"; mCachedDiary.reset(); }
 
 
     /**

--- a/src/ElevenLabsClient.cpp
+++ b/src/ElevenLabsClient.cpp
@@ -17,6 +17,7 @@ AJSON_FIELDS(ElevenLabsClient::TextToSpeechRequest,
 
 AFuture<ElevenLabsClient::TextToSpeechResponse>
 ElevenLabsClient::textToSpeech(const TextToSpeechRequest& request) {
+    ALOG_TRACE(LOG_TAG) << "textToSpeech";
     if (baseUrl.empty()) {
         throw AException("ElevenLabs endpoint URL not configured");
     }

--- a/src/ImageGenerator.cpp
+++ b/src/ImageGenerator.cpp
@@ -29,6 +29,7 @@ static AJson parseResponse(AString content) {
 }
 
 AFuture<ImageGenerator::GalleryImage> ImageGenerator::generate(AString description) {
+    ALOG_TRACE(LOG_TAG) << "generate: " << description;
     AString appearancePrompt = kuni_character::getAppearancePrompt();
     int trialIndex = 0;
 
@@ -136,6 +137,7 @@ static constexpr auto LOL_WHAT = {
 };
 
 AFuture<> ImageGenerator::engineerPrompt(PromptPair& out, const AString& description, const AString& appearancePrompt, const AString& feedback) {
+    ALOG_TRACE(LOG_TAG) << "engineerPrompt description=" << description << " appearancePrompt=" << appearancePrompt << " feedback=" << feedback;
     auto safeDescription = description;
     for (const auto& word : LOL_WHAT) {
         safeDescription.replaceAll(word, "");
@@ -261,6 +263,7 @@ Negative prompt is what to avoid in the image.
 }
 
 AFuture<ImageGenerator::AssessmentResult> ImageGenerator::assessImage(const AImage& image, const AString& description) {
+    ALOG_TRACE(LOG_TAG) << "assessImage description=" << description;
     OpenAIChat chat = mChatClient;
     // Note: mChatClient.config should ideally be a vision-capable model.
     chat.systemPrompt = R"(

--- a/src/KuniCharacter.cpp
+++ b/src/KuniCharacter.cpp
@@ -4,6 +4,7 @@
 
 #include "KuniCharacter.h"
 
+#include "AUI/Logging/ALogger.h"
 #include "AUI/Common/AByteBuffer.h"
 #include "AUI/IO/AFileInputStream.h"
 #include "AUI/IO/AFileOutputStream.h"
@@ -35,6 +36,7 @@ static AString readOrCreate(AStringView fileName, AStringView defaultContent)
 
 AString kuni_character::getBasePrompt()
 {
+    ALOG_TRACE("KuniCharacter") << "getBasePrompt";
     return readOrCreate("character_base.md", R"(---
 This is system prompt of the LLM character. You can adjust it for your taste (i.e., change name).
 In this prompt we define character name, some background and. We inspire LLM into thinking they have thoughts, emotions
@@ -159,6 +161,7 @@ Say hi. I'm usually around.
 
 AString kuni_character::getAppearancePrompt()
 {
+    ALOG_TRACE("KuniCharacter") << "getAppearancePrompt";
     return readOrCreate("character_appearance.md", R"(---
 Visual character description. This includes freeform description (of the profile photo) and a stable diffusion prompt.
 Both prompts are included into system prompt.

--- a/src/OpenAIChat.cpp
+++ b/src/OpenAIChat.cpp
@@ -56,6 +56,7 @@ AJSON_FIELDS(StreamingResponse::Choice,
     )
 
 AFuture<OpenAIChat::Response> OpenAIChat::chat(AString message) {
+    ALOG_TRACE(LOG_TAG) << "chat(" << message << ")";
     return chat({
         {Message::Role::USER, std::move(message)},
     });
@@ -115,6 +116,7 @@ struct AJsonConv<AVector<OpenAIChat::Message>> {
 
 
 AString OpenAIChat::embedImage(AImageView image) {
+    ALOG_TRACE(LOG_TAG) << "embedImage";
     AByteBuffer jpg;
     auto resized = image.resizedLinearDownscale({672, 672});
     JpgImageLoader::save(jpg, resized);
@@ -123,6 +125,7 @@ AString OpenAIChat::embedImage(AImageView image) {
 }
 
 AJson OpenAIChat::makeQueryString(AVector<OpenAIChat::Message> messages) {
+    ALOG_TRACE(LOG_TAG) << "makeQueryString";
     AJson json {
         {
           "messages",
@@ -274,6 +277,7 @@ _<OpenAIChat::StreamingResponse> OpenAIChat::chatStreaming(AVector<Message> mess
 }
 
 AFuture<std::valarray<double>> OpenAIChat::embedding(AString input) {
+    ALOG_TRACE(LOG_TAG) << "embedding";
     if (input.empty()) {
         input = " ";
     }

--- a/src/OpenAITools.cpp
+++ b/src/OpenAITools.cpp
@@ -18,6 +18,7 @@ AJSON_FIELDS(OpenAITools::Tool, AJSON_FIELDS_ENTRY(type) AJSON_FIELDS_ENTRY(name
 
 
 OpenAITools::OpenAITools(std::initializer_list<Tool> tools) {
+    ALOG_TRACE("OpenAITools") << "OpenAITools";
     for (const auto& tool : tools) {
         AUI_ASSERT(tool.handler != nullptr);
         mHandlers[tool.name] = tool;
@@ -25,6 +26,7 @@ OpenAITools::OpenAITools(std::initializer_list<Tool> tools) {
 }
 
 AFuture<AVector<OpenAIChat::Message>> OpenAITools::handleToolCalls(const AVector<OpenAIChat::Message::ToolCall>& toolCalls) {
+    ALOG_TRACE("OpenAITools") << "handleToolCalls";
     AVector<OpenAIChat::Message> result;
     for (const auto& toolCall : toolCalls) {
         result << OpenAIChat::Message{
@@ -48,6 +50,7 @@ AFuture<AVector<OpenAIChat::Message>> OpenAITools::handleToolCalls(const AVector
 }
 
 AJson OpenAITools::asJson() const {
+    ALOG_TRACE("OpenAITools") << "asJson";
     return ranges::view::transform(mHandlers, [](const auto& tool) {
         return AJson::Object{
             {"type", tool.second.type },

--- a/src/StableDiffusionClient.cpp
+++ b/src/StableDiffusionClient.cpp
@@ -36,6 +36,7 @@ AJSON_FIELDS(StableDiffusionClient::Txt2ImgRequest,
              AJSON_FIELDS_ENTRY(denoising_strength))
 
 AFuture<StableDiffusionClient::Txt2ImgResponse> StableDiffusionClient::txt2img(const Txt2ImgRequest& request) {
+    ALOG_TRACE(LOG_TAG) << "txt2img";
     auto query = AJson::toString(aui::to_json(request));
     ALOG_TRACE(LOG_TAG) << "Query: " << query;
 

--- a/src/WebSearch.cpp
+++ b/src/WebSearch.cpp
@@ -20,7 +20,7 @@ AJSON_FIELDS(web::Result,
     )
 
 AFuture<AVector<web::Result>> web::search(AString query, int maxResults) {
-    ALOG_DEBUG(LOG_TAG) << "search query=\"" << query << "\"";
+    ALOG_TRACE(LOG_TAG) << "web::search: " << query;
     // Build JSON body
     AJson body = AJson::Object{{"query", std::move(query)} };
     if (maxResults> 0) {
@@ -44,7 +44,7 @@ AFuture<AVector<web::Result>> web::search(AString query, int maxResults) {
 }
 
 AFuture<AString> web::searchAI(AString query) {
-    ALOG_DEBUG(LOG_TAG) << "searchAI query=\"" << query << "\"";
+    ALOG_TRACE(LOG_TAG) << "web::searchAI: " << query;
     OpenAITools tools {
         OpenAITools::Tool {
             .name = "query",

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -44,6 +44,7 @@ namespace {
     class App : public AppBase {
     public:
         App(): AppBase("data") {
+            ALOG_TRACE(LOG_TAG) << "App::App";
             mTelegram->onEvent = [this](td::td_api::object_ptr<td::td_api::Object> event) {
                 td::td_api::downcast_call(*event,
                                           [this](auto& u) { mAsync << this->handleTelegramEvent(std::move(u)); });
@@ -55,6 +56,7 @@ namespace {
 
     protected:
         AFuture<> telegramPostMessage(int64_t chatId, AString text, AOptional<_<AImage>> photo = std::nullopt, AOptional<APath> audioPath = std::nullopt, int64_t replyTo = 0) {
+            ALOG_TRACE(LOG_TAG) << "telegramPostMessage: chat_id" << chatId << " text=" << text << " photo=" << photo << " audioPath=" << audioPath << " replyTo=" << replyTo;
             // Check lockdown mode - only allow PAPIK_CHAT_ID if lockdown is enabled
             if constexpr (config::LOCKDOWN_MODE) {
                 if (chatId != config::PAPIK_CHAT_ID) {

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -43,7 +43,7 @@ namespace {
 
     class App : public AppBase {
     public:
-        App(): AppBase("data") {
+        App(_<TelegramClient> telegram): AppBase("data"), mTelegram(std::move(telegram)) {
             ALOG_TRACE(LOG_TAG) << "App::App";
             mTelegram->onEvent = [this](td::td_api::object_ptr<td::td_api::Object> event) {
                 td::td_api::downcast_call(*event,
@@ -356,7 +356,7 @@ Use absolute time in your queries.
         }
 
     private:
-        _<TelegramClient> mTelegram = _new<TelegramClient>();
+        _<TelegramClient> mTelegram;
 
         [[nodiscard]]
         AFuture<> llmuiFormatChatList(AString& result, std::span<td::td_api::object_ptr<td::td_api::chat>> chats) {
@@ -1435,29 +1435,35 @@ AUI_ENTRY {
     }
 
     using namespace std::chrono_literals;
-    auto app = _new<App>();
-
-    _new<AThread>([] {
-        std::cin.get();
-        gEventLoop.stop();
-    })->start();
+    auto telegram = _new<TelegramClient>();
 
     AAsyncHolder async;
-    async << [](_<App> app) -> AFuture<> {
+    async << [](_<TelegramClient> telegram) -> AFuture<> {
         ALogger::info(LOG_TAG) << "Waiting for Telegram network...";
-        co_await app->telegram()->waitForConnection();
+        co_await telegram->waitForConnection();
         ALogger::info(LOG_TAG) << "Connected to Telegram";
-
         // app->actProactively(); // for tests
-    }(app);
+    }(telegram);
+
+    _<App> app;
+    AObject::connect(telegram->loggedIn, telegram, [&] {
+        app = _new<App>(telegram);
+        _new<AThread>([] {
+            ALogger::info(LOG_TAG) << "Bot is up and running. Press enter to shutdown gracefully.";
+            std::cin.get();
+            ALogger::info(LOG_TAG) << "Bot is shutting down. Please give some time to dump remaining context";
+            gEventLoop.stop();
+        })->start();
+    });
 
     IEventLoop::Handle h(&gEventLoop);
     gEventLoop.loop();
 
-    ALogger::info(LOG_TAG) << "Bot is shutting down. Please give some time to dump remaining context";
-    auto d = app->diaryDumpMessages();
-    while (!d.hasResult()) {
-        AThread::processMessages();
+    if (app) {
+        auto d = app->diaryDumpMessages();
+        while (!d.hasResult()) {
+            AThread::processMessages();
+        }
     }
 
     return 0;

--- a/src/telegram/TelegramClient.cpp
+++ b/src/telegram/TelegramClient.cpp
@@ -16,6 +16,7 @@ namespace {
 
 
 TelegramClient::TelegramClient() : mTgUpdateTimer(_new<ATimer>(1s)) {
+    ALOG_TRACE(LOG_TAG) << "TelegramClient::TelegramClient";
     setSlotsCallsOnlyOnMyThread(true);
 
     td::ClientManager::execute(td::td_api::make_object<td::td_api::setLogVerbosityLevel>(1));
@@ -26,6 +27,7 @@ TelegramClient::TelegramClient() : mTgUpdateTimer(_new<ATimer>(1s)) {
 }
 
 AFuture<TelegramClient::Object> TelegramClient::sendQuery(td::td_api::object_ptr<td::td_api::Function> f) {
+    ALOG_TRACE(LOG_TAG) << "sendQuery";
     if (mQueryCountLastUpdate++ >= 5) {
         // Telegram is strict about using 3rdparty telegram clients. For this reason, we have to ensure that we wouldn't
         // trigger their security leading to ban of the account.
@@ -42,6 +44,7 @@ AFuture<TelegramClient::Object> TelegramClient::sendQuery(td::td_api::object_ptr
 }
 
 void TelegramClient::initClientManager() {
+    ALOG_TRACE(LOG_TAG) << "initClientManager";
     mClientManager = std::make_unique<td::ClientManager>();
     mClientId = mClientManager->create_client_id();
     sendQueryWithResult(td::td_api::make_object<td::td_api::getOption>("version"))
@@ -56,6 +59,7 @@ void TelegramClient::initClientManager() {
 }
 
 void TelegramClient::update() {
+    ALOG_TRACE(LOG_TAG) << "update";
     mQueryCountLastUpdate = 0;
     for (;;) {
         auto response = mClientManager->receive(0);
@@ -67,6 +71,7 @@ void TelegramClient::update() {
 }
 
 void TelegramClient::processResponse(td::ClientManager::Response response) {
+    ALOG_TRACE(LOG_TAG) << "processResponse";
     if (!response.object) {
         return;
     }
@@ -82,6 +87,7 @@ void TelegramClient::processResponse(td::ClientManager::Response response) {
 }
 
 void TelegramClient::commonHandler(td::tl::unique_ptr<td::td_api::Object> object) {
+    ALOG_TRACE(LOG_TAG) << "commonHandler";
     td::td_api::downcast_call(
         *object,
         aui::lambda_overloaded{

--- a/src/telegram/TelegramClient.cpp
+++ b/src/telegram/TelegramClient.cpp
@@ -108,7 +108,10 @@ void TelegramClient::commonHandler(td::tl::unique_ptr<td::td_api::Object> object
                             parameters->application_version_ = AUI_PP_STRINGIZE(AUI_CMAKE_PROJECT_VERSION);
                             sendQuery(std::move(parameters));
                         },
-                        [this](td::td_api::authorizationStateReady& u) {},
+                        [this](td::td_api::authorizationStateReady& u) {
+                            ALogger::info(LOG_TAG) << "[Authentication] logged in.";
+                            emit loggedIn;
+                        },
                         [this](td::td_api::authorizationStateWaitPhoneNumber& s) {
                             ALogger::info(LOG_TAG) << "[Authentication] required. Please supply phone number to stdin";
 

--- a/src/telegram/TelegramClient.h
+++ b/src/telegram/TelegramClient.h
@@ -6,6 +6,7 @@
 #include <td/telegram/td_api.hpp>
 
 #include "AUI/Common/AMap.h"
+#include "AUI/Common/ASignal.h"
 #include "AUI/Thread/AAsyncHolder.h"
 #include "AUI/Thread/AFuture.h"
 
@@ -45,6 +46,8 @@ public:
     static td::td_api::object_ptr<T> toPtr(T&& t) {
         return td::td_api::make_object<T>(std::forward<T>(t));
     }
+
+    emits<> loggedIn;
 
 private:
     AFuture<> mWaitForConnection;

